### DISCUSSION TITLE
Some ideas about DoH and customizing it per user

### DISCUSSION THREAD

**Discussion by robin98** [hiddify/Hiddify-Manager#152] [Ideas ایدهها]
در حالت فعلی هر درخواستی که به DoH هایدیفای می رسد، ظاهرا توسط DNS گوگل پاسخ داده می شود. بهتر است که:

- در مستندات، روش تغییر پیشفرض را از گوگل به هر سرویس دهنده دیگری شرح بدهید.
- امکانی در پنل وب تعبیه شود که ادمین بتواند پاسخ دهنده سرویس DoH سرورش را از بین چند سرویس دهنده و حتی یک سرویس دهنده کاستوم (دقیقا مشابه فیلدی که در کروم وجود دارد) انتخاب یا وارد کند.

### ایده جدید: به ازای هر یوزری که ساخته می شود، امکان تعیین یک تحلیل کننده متفاوت مهیا باشد.
به طور مثال هنگامی که یوزر X با کانفیگ های شخصی اش تحت بستر مثلا Vless یا Trojan متصل می شود، تحلیل درخواست های DNS او که به سمت سرور می رسد با گوگل باشد.
اما در مورد یوزر Y، تحلیل درخواست های DNS او مثلا بر عهده quad9 باشد و الی آخر... .
نکته: در حال حاضر، وظیفه تحلیل درخواست های DNS تمام کاربران بوسیله دیفالت سرور انجام می شود.

استفاده های این قابلیت:

- امکان تنظیم سرویس های ادبلاکر یا بلاک کردن فهرستی از دامنه ها یا سرویس ها برای یک کانفیگ مشخص بدون تداخل با سایر یوزرها
- امکان فعالسازی نظارت والدین روی اکانت هایی که در دسترس کودکان قرار داده می شوند با تنظیم تحلیلگر آن یوزر روی DNS های فمیلی مثلا 1.1.1.3 در کلادفلیر
- امکان استفاده از قابلیت های وسیعی که سرویس دهنده های DNS های امن همچون NextDNS یا Controld که در اختیار مصرف کنندگان می گذارند

در صورت پیاده سازی این مدل، رنج وسیعی از امکانات با کمک این ایده قابل اجرا هست بدون آنکه زحمت توسعه یا کدنویسی تک تک آنها روی دوش هایدیفای باشد.

در مورد روش پیاده سازی این مدل اما، ایده ای ندارم. یک ایده شاید این باشد که چند Instance همزمان از کل هایدیفای با کمک داکر پیاده سازی شود و با روشی به هر Container یک Resolver مجزا اختصاص داد؛ که البته روش بهینه و قابل قبولی نیست.

در سیستم عامل لینوکس ابونتو نسخه 20 و 22 با رجوع به فایل زیر:
`/etc/systemd/resolved.conf`
می توان با وارد کردن مقادیری به شکل زیر:
```
[Resolve]
DNS=45.90.28.0#$USER_ID.dns.nextdns.io
DNS=2a07:a8c0::#$USER_ID.dns.nextdns.io
DNS=45.90.30.0#$USER_ID.dns.nextdns.io
DNS=2a07:a8c1::#$USER_ID.dns.nextdns.io
DNSOverTLS=yes
```
ترافیک درخواست های DNS کل سیستم عامل را روی یکی از دو پروتکل tcp یا همان حالت رایج، و یا DoT تنظیم کرد.

**Comment by hiddify-com** [comment]
ایده جالبی هست اما امکان شخصی سازی اون نیست برای هر یوزر
میشه برای همه کاربران تغییر داد ولی

**Comment by robin98** [comment]
بوسیله اون چند خط آخر اگه انجام بشه، بله برای کل سیستم عامل تغییر پیدا می کنه.

از اونجایی که درک عمیقی از لینوکس یا چگونگی کارکرد سرویس های xray ندارم، نمیدونم که پیاده سازی همچین ایده ای نیاز به چه پیش نیازهایی داره. شاید اگر نظر چندنفر از توسعه دهندگان پروژه های v2ray و xray رو در گیت هاب بپرسیم به اطلاعات دقیق تری برسیم، یا شاید حتی همین الان هم این ویژگی توسط آنها پیاده سازی شده باشه.

به هر حال با چیزی که من به ذهنم رسید و جستجو کردم، از سرویس chatgpt جواب زیر رو گرفتم که ایده ای در مورد صحت و عملی بودنش ندارم؛ همچنین من نمیدونم که چطور میشه بین یوزرها تفاوت قائل شد، از روی پورت لوکال؟ از روی PID؟ یا هر چیز دیگر… .
در سوالی که من پرسیدم، هر یوزر رو یک process مجزا در نظر گرفتم.

### Question:

> I have a Linux server with ubuntu 22 server installed on it. In the other hand, I have two different DNS server address in DoT protocol. Now I want to set both of this DNS server addresses on my Linux server and define a rule likes: If service A want to resolve a domain name, it must use the resolver A, and if service B want to resolve a domain it must use the resolver B.

### Answer:

> You can configure your Ubuntu 22 server to use different DNS resolvers for different services by modifying the DNS resolver settings in the relevant configuration files for each service. Here are the general steps you can follow:
> 
> Check which DNS resolver your server is currently using by running the following command:
> 
> 
> ```
> systemd-resolve --status
> ```
> This will display the current DNS resolver and related settings.
> 
> Edit the `/etc/systemd/resolved.conf` file to add the DNS server addresses you want to use. Add the following lines to the file:
> 
> 
> ```
> [Resolve]
> DNS=1.1.1.1 2.2.2.2
> ```
> Replace `1.1.1.1` and `2.2.2.2` with the IP addresses of the two DNS servers you want to use.
> 
> Restart the systemd-resolved service to apply the changes:
> 
> 
> `systemctl restart systemd-resolved.service`
> For each service you want to configure, create a new .service file in the `/etc/systemd/system/ directory`. For example, to configure Service A, create a file called `/etc/systemd/system/service-a.service`.
> 
> In the new .service file, add the following lines to configure the DNS resolver for Service A:
> 
> 
> ```
> [Service]
> Environment="DNS_SERVER=1.1.1.1"
> ```
> Replace `1.1.1.1` with the IP address of the DNS resolver you want Service A to use.
> 
> Repeat steps 4-6 for Service B, but use a different file name and IP address for the DNS resolver.
> 
> Restart each service for the changes to take effect.
> 
> By setting the `DNS_SERVER` environment variable for each service, you can specify which DNS resolver that service should use. Note that not all services may support environment variables for DNS resolver settings, so you may need to check the documentation for each service to determine the appropriate configuration method.

شاید بعد اینهمه، پیاده سازی این سرویس نشدنی هم نباشه.
متاسفانه همانطور که اشاره کردم دانش خاصی در زمینه لینوکس ندارم و برنامه نویسی را در حد پرینت Hello World (با کمک گرفتن از Google) بلد هستم، در غیراینصورت حتما در پیاده سازی این ایده مشارکت می کردم.

**Comment by raminrez** [comment]
@hiddify-com  @lymanjre 
سلام دوستان
آپدیتی درمورد چگونگی استفاده از DOH و تغییر پروایدر پیشفرض دارید؟
